File: src/game/creature/ZombieTurret.java
package game.creature;
import static org.lwjgl.opengl.GL11.*;
import game.TextureManager;
import game.Texture;
import game.Vector;
import game.level.Level;


public class ZombieTurret extends Creature {
	boolean detectedPlayer = false;
	float headAngle;
	
	public ZombieTurret() {
		super(10, TextureManager.TURRETBODY);
		drawRadius *= 2;
		hitRadius = drawRadius - 0.1f;
		mass = 100;
	}

	@Override
	public void update(float dt, Level level) {
		if (!detectedPlayer && !level.tileGrid.hitsWall(pos, level.player.pos) && level.player.pos.distance(pos) <= 10) {
			detectedPlayer = true;
		}
		if (detectedPlayer) {
			Vector target = level.player.pos;
			
			Vector facing = new Vector(headAngle);
			int dir;
			if (target.sub(pos).dot(facing.perpCCW()) > 0) dir = 1;
			else dir = -1;
			
			float speed = 2*(1 - target.sub(pos).direction().dot(facing) + 1);
			headAngle += speed*dir*dt;
			
		} else {
			headAngle = angle;
		}
		
		super.update(dt, level);
	}
	
	@Override
	public void render() {
		super.render();
		
		glPushMatrix();
		glTranslatef(pos.x, pos.y, 0);
		glRotatef((float) (headAngle*180/Math.PI), 0, 0, 1);
		float headRadius = 0.5f;
		glScalef(headRadius, headRadius, 1);
		
		TextureManager.ZOMBIEHEAD.bind();
		glColor4f(1, 1, 1, invincible > 0 ? 0.5f : 1);
		glBegin(GL_QUADS);
		glTexCoord2f(0, 1);
		glVertex2f(-1, -1);
		glTexCoord2f(1, 1);
		glVertex2f(1, -1);
		glTexCoord2f(1, 0);
		glVertex2f(1, 1);
		glTexCoord2f(0, 0);
		glVertex2f(-1, 1);
		glEnd();
                Texture.bindNone();
		
		glPopMatrix();
	}
}

File: ts/game/creature/ZombieTurret.ts
import * as PIXI from 'pixi.js';
import { TextureManager } from '../TextureManager.js';
import { Vector } from '../Vector.js';
import { Level } from '../level/Level.js';
import { Creature } from './Creature.js';

export class ZombieTurret extends Creature {
  detectedPlayer = false;
  headAngle = 0;
  private headSprite: PIXI.Sprite;

  constructor() {
    super(10, TextureManager.TURRETBODY);
    this.drawRadius *= 2;
    this.hitRadius = this.drawRadius - 0.1;
    this.mass = 100;
    this.headSprite = new PIXI.Sprite(TextureManager.ZOMBIEHEAD);
    this.headSprite.anchor.set(0.5);
  }

  update(dt: number, level: Level) {
    if (!this.detectedPlayer && !level.tileGrid.hitsWall(this.pos, level.player.pos) && level.player.pos.distance(this.pos) <= 10) {
      this.detectedPlayer = true;
    }
    if (this.detectedPlayer) {
      const target = level.player.pos;
      const facing = new Vector(this.headAngle);
      const dir = target.sub(this.pos).dot(facing.perpCCW()) > 0 ? 1 : -1;
      const speed = 2 * (1 - target.sub(this.pos).direction().dot(facing) + 1);
      this.headAngle += speed * dir * dt;
    } else {
      this.headAngle = this.angle;
    }
    super.update(dt, level);
  }

  render(container: PIXI.Container) {
    super.render(container);
    if (!this.headSprite.parent) {
      container.addChild(this.headSprite);
      const headRadius = 0.5;
      this.headSprite.width = headRadius * 2;
      this.headSprite.height = headRadius * 2;
      this.headSprite.scale.y *= -1;
    }
    this.headSprite.position.set(this.pos.x, this.pos.y);
    this.headSprite.rotation = this.headAngle;
    this.headSprite.alpha = this.invincible > 0 ? 0.5 : 1;
  }
}
